account_invoice_change_currency/models/account_change_currency.py

```diff
@@ -0,0 +1,311 @@
+FROM mail_tracking_value as mtv INNER JOIN mail_message AS mm
+ON mtv.mail_message_id = mm.id INNER JOIN
+ir_model_fields AS field
+ON field.id = mtv.field """
+        params = (self._name, tuple(self.message_ids.ids))
+        if skip_update_currency:
+            query += " AND mm.subtype_id != %s "
+            params = (subtype_id.id, self._name, tuple(self.message_ids.ids))
+        query += """ WHERE field.name = 'currency_id' AND
+field.model = %s AND mtv.mail_message_id IN %s
+ORDER BY mtv.write_date DESC, mtv.id DESC LIMIT 1"""
+        self.env.cr.execute(query, params)
+        res = self.env.cr.dictfetchone()
+        value = False
+        if res:
+            value = res["old_value_integer"]
+            if res["subtype_id"] == subtype_create_id.id:
+                value = res["new_value_integer"]
+        return self.currency_id.browse(value)
+
+    def get_last_rate(self):
+        self.ensure_one()
+        subtype_create_id = self.env.ref("account.mt_invoice_created")
+        ir_model_fields = self.env["ir.model.fields"]
+        ir_model_fields_ids = (
+            ir_model_fields.sudo()
+            .search(
+                [
+                    ("name", "in", ["rate", "currency_id"]),
+                    ("model", "in", [self._name, "res.currency"]),
+                ]
+            )
+            .ids
+        )
+        last_values = (
+            self.env["mail.tracking.value"]
+            .sudo()
+            .search(
+                [
+                    ("mail_message_id", "in", self.message_ids.ids),
+                    ("field", "in", ir_model_fields_ids),
+                ],
+                limit=2,
+                order="write_date desc, id desc",
+            )
+        )
+        if len(last_values) == 2:
+            currency_value, rate_value = sorted(
+                last_values, key=lambda r: r.field.id, reverse=True
+            )
+            return (
+                self.currency_id.browse(currency_value.new_value_integer),
+                rate_value.new_value_float,
+            )
+        if (
+            len(last_values) == 1
+            and last_values.mail_message_id.subtype_id == subtype_create_id
+        ):
+            return (self.currency_id.browse(last_values.new_value_integer), None)
+        return self.currency_id.browse(None), None
+
+    def _toggle_forced_rate(self):
+        self.ensure_one()
+        state = self.get_force_rate_state()
+        if not state:
+            self._track_force_rate()
+        else:
+            force = not state.new_value_integer
+            self._track_force_rate(force)
+
+    def get_force_rate_state(self):
+        self.ensure_one()
+        subtype_id = self.env.ref("account_invoice_change_currency.mt_force_rate")
+        ir_model_fields = self.env["ir.model.fields"]
+        ir_model_fields_id = (
+            ir_model_fields.sudo()
+            .search([("name", "=", "to_check"), ("model", "=", self._name)])
```

Comment:
`ir.model.fields` has a method `get(model_name, field_name)` that does this for you

Reply:
The change that I applied makes this code unnecessary